psa-crypto/src/operations/cipher.rs

```diff
@@ -2,3 +2,146 @@
 // SPDX-License-Identifier: Apache-2.0
 
 //! # Unauthenticated Ciphers operations
+
+use crate::initialized;
+use crate::types::algorithm::Cipher;
+use crate::types::key::Id;
+use crate::types::status::{Result, Status};
+
+fn crypt(
+    encrypt: bool,
+    key_id: Id,
+    alg: Cipher,
+    plaintext: &[u8],
+    iv: &[u8],
+    ciphertext: &mut [u8],
+) -> Result<usize> {
+    initialized()?;
+
+    let mut operation: psa_crypto_sys::psa_cipher_operation_t =
+        unsafe { psa_crypto_sys::psa_cipher_operation_init() };
+
+    Status::from(unsafe {
+        (if encrypt {
+            psa_crypto_sys::psa_cipher_encrypt_setup
+        } else {
+            psa_crypto_sys::psa_cipher_decrypt_setup
+        })(&mut operation, key_id.0, alg.into())
+    })
+    .to_result()?;
+
+    Status::from(unsafe {
+        psa_crypto_sys::psa_cipher_set_iv(&mut operation, iv.as_ptr(), iv.len())
+    })
+    .to_result()?;
+
+    let mut output_length = 0;
+    Status::from(unsafe {
+        psa_crypto_sys::psa_cipher_update(
+            &mut operation,
+            plaintext.as_ptr(),
+            plaintext.len(),
+            ciphertext.as_mut_ptr(),
+            ciphertext.len(),
+            &mut output_length,
+        )
+    })
+    .to_result()?;
+
+    let mut output_length_finish = 0;
+    Status::from(unsafe {
+        psa_crypto_sys::psa_cipher_finish(
+            &mut operation,
+            ciphertext.as_mut_ptr().add(output_length),
+            ciphertext.len() - output_length,
+            &mut output_length_finish,
+        )
+    })
+    .to_result()?;
+
+    Ok(output_length + output_length_finish)
+}
+
+/// Encrypt a short message with a key
+///
+/// The encrypted message is written in `ciphertext`. The function returns the number of bytes written.
+///
+/// # Example
+///
+/// ```
+/// # use psa_crypto::operations::cipher::encrypt;
+/// # use psa_crypto::operations::key_management::generate;
+/// # use psa_crypto::types::algorithm::Cipher;
+/// # use psa_crypto::types::key::{Attributes, Type, Lifetime, Policy, UsageFlags};
+/// #
+/// # let mut usage_flags: UsageFlags = Default::default();
+/// # usage_flags.set_encrypt();
+/// # let mut attributes = Attributes {
+/// #     key_type: Type::Aes,
+/// #     bits: 128,
+/// #     lifetime: Lifetime::Volatile,
+/// #     policy: Policy {
+/// #         usage_flags,
+/// #         permitted_algorithms: Cipher::CbcNoPadding.into(),
+/// #     },
+/// # };
+/// # const MESSAGE: [u8; 16] = [0; 16];
+/// psa_crypto::init().unwrap();
+/// let my_key = generate(attributes, None).unwrap();
+/// let alg = Cipher::CbcNoPadding;
+/// let iv = vec![0; 16];
+/// let mut encrypted_message = vec![0; MESSAGE.len()];
+///
+/// let size = encrypt(my_key, alg, &MESSAGE, &iv, &mut encrypted_message).unwrap();
+/// ```
+pub fn encrypt(
+    key_id: Id,
+    alg: Cipher,
+    plaintext: &[u8],
+    iv: &[u8],
```

Comment:
Could you make the `iv` field optional for encryption, and call `psa_cipher_generate_iv` if it's not provided?

Reply:
> Could you make the `iv` field optional for encryption, and call `psa_cipher_generate_iv` if it's not provided?

I don't think that makes sense with any of these algorithms because the IV isn't included in the encrypted message so a message generated with a random IV that isn't revealed to the caller has no feasible use.

Reply:
Ah, I was under the impression the generated IV would be prepended to the ciphertext. Hmmm, I would still prefer if the IV could be left to the implementation to handle, especially since it's the recommended approach in the spec. Will try to think of a way

Reply:
Hmm, perhaps we can leave as is for now and add another method, something like `encrypt_with_new_iv`, when there's a need.